src/oci.rs

```diff
@@ -163,24 +168,48 @@ fn append_dir_all_with_xattrs(
                 let mtime = filetime::FileTime::from_unix_time(clamp_mtime, 0);
                 filetime::set_symlink_file_times(entry.path(), mtime, mtime)?;
             }
-            add_pax_extension_header(entry.path(), tar)?;
-            tar.append_path_with_name(entry.path(), rel_path)?;
+            add_pax_extension_header(entry.path(), builder)?;
+            builder.append_path_with_name(entry.path(), rel_path)?;
         } else if entry.file_type().is_file() || entry.file_type().is_dir() {
-            add_pax_extension_header(entry.path(), tar)?;
+            add_pax_extension_header(entry.path(), builder)?;
+
+            // If this is a hardlink, add a link header instead of the file
+            // if this isn't the first time we've seen this inode
+            if meta.nlink() > 1 {
+                match hardlinks.entry((meta.dev(), meta.ino())) {
+                    Entry::Occupied(e) => {
+                        // Add link header and continue to next entry
+                        let mut header = tar::Header::new_gnu();
+                        header.set_metadata(&meta);
+                        if meta.mtime() > clamp_mtime {
+                            header.set_mtime(clamp_mtime as u64);
+                        }
+                        header.set_entry_type(tar::EntryType::Link);
+                        header.set_cksum();
+                        builder.append_link(&mut header, &rel_path, e.get())?;
+                        continue;
```

Comment:
Happy with this implementation. I do think you should consider refactoring this else block into modular functions for testability. Additionally refactoring would remove the need for `continue`.

Reply:
if i add more fine grained testing then I will do